hazelcast/src/main/java/com/hazelcast/map/impl/MapIndexSynchronizer.java

```diff
@@ -0,0 +1,118 @@
+ */
+class MapIndexSynchronizer {
+
+    protected MapServiceContext mapServiceContext;
+    protected OperationService operationService;
+    protected SerializationService serializationService;
+    protected NodeEngine nodeEngine;
+
+    private ILogger logger;
+    private Version currentVersion;
+
+    public MapIndexSynchronizer(MapServiceContext mapServiceContext, NodeEngine nodeEngine) {
+        this.mapServiceContext = mapServiceContext;
+        this.operationService = nodeEngine.getOperationService();
+        this.serializationService = nodeEngine.getSerializationService();
+        this.nodeEngine = nodeEngine;
+        this.logger = nodeEngine.getLogger(getClass());
+    }
+
+    public void onClusterVersionChange(Version newVersion) {
+        if (isV38toV39transition(newVersion)) {
+            synchronizeIndexes();
+        }
+        currentVersion = newVersion;
+    }
+
+    private void synchronizeIndexes() {
+        logger.info("Running MapIndexSynchronizer");
+        List<MapIndexInfo> mapIndexInfos = getIndexesToSynchronize();
+        if (!mapIndexInfos.isEmpty()) {
+            executeIndexSync(mapIndexInfos);
+        }
+    }
+
+    private List<MapIndexInfo> getIndexesToSynchronize() {
+        List<MapIndexInfo> mapIndexInfos = new ArrayList<MapIndexInfo>();
+        for (Map.Entry<String, MapContainer> entry : mapServiceContext.getMapContainers().entrySet()) {
+            MapContainer mapContainer = entry.getValue();
+            int indexesToSynchronize = mapContainer.getPartitionIndexesToAdd().size();
+            if (indexesToSynchronize > 0) {
+                String mapName = entry.getKey();
+                logger.info("Scheduling " + indexesToSynchronize + " indexes sync for map " + mapName);
+                MapIndexInfo mapIndexInfo = new MapIndexInfo(mapName);
+                for (IndexInfo indexInfo : mapContainer.getPartitionIndexesToAdd()) {
+                    mapIndexInfo.addIndexInfo(indexInfo.getAttributeName(), indexInfo.isOrdered());
+                }
+                mapIndexInfos.add(mapIndexInfo);
+            }
+            mapContainer.clearPartitionIndexesToAdd();
+        }
+        return mapIndexInfos;
+    }
+
+    private void executeIndexSync(List<MapIndexInfo> mapIndexInfos) {
+        OperationServiceImpl operationServiceImpl = (OperationServiceImpl) operationService;
+        for (int partitionId : mapServiceContext.getOwnedPartitions()) {
+            SynchronizeIndexesForPartitionTask task = new SynchronizeIndexesForPartitionTask(partitionId, mapIndexInfos,
+                    mapServiceContext.getService(), serializationService,
+                    (InternalPartitionService) nodeEngine.getPartitionService());
+            operationServiceImpl.execute(task);
+        }
+    }
+
+    private boolean isV38toV39transition(Version newVersion) {
+        return currentVersion != null && currentVersion.equals(Versions.V3_8) && newVersion.equals(Versions.V3_9);
```

Comment:
when is the `currentVersion` null?

Reply:
it's just the synchronizer. It gets a valid version on the first call.